nni/runtime/tuner_command_channel/websocket.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+"""
+Synchronized and object-oriented WebSocket class.
+
+WebSocket guarantees that messages will not be divided at API level.
+"""
+
+from __future__ import annotations
+
+__all__ = ['WebSocket']
+
+import asyncio
+import logging
+from threading import Lock, Thread
+from typing import Any
+
+import websockets
+
+_logger = logging.getLogger(__name__)
+
+# the singleton event loop
+_event_loop: asyncio.AbstractEventLoop = None  # type: ignore
```

Comment:
why `_event_loop` should be a singleton？

Reply:
Because that is how asyncio designed to work.
In fact there should be only one global event loop thread, not one per module. But since nni does not use asyncio elsewhere (except trial keeper), for simplicity we put the event loop in websocket module for now.